File: scripts/render_agents.py
#!/usr/bin/env python3
from __future__ import annotations

import re
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
TEMPLATE = ROOT / "AGENTS.template.md"
OUTPUT = ROOT / "GLOBAL_AGENTS.md"
RULES_DIR = ROOT / "rules"
BEGIN = "<!-- BEGIN GENERATED RULES -->"
END = "<!-- END GENERATED RULES -->"


def parse_rule(path: Path) -> tuple[str, str]:
    text = path.read_text(encoding="utf-8")
    description = ""
    body = text

    if text.startswith("---\n"):
        _, frontmatter, body = text.split("---\n", 2)
        for line in frontmatter.splitlines():
            key, sep, value = line.partition(":")
            if sep and key.strip() in {"description", "summary"}:
                description = value.strip().strip("\"'")
                break

    title = ""
    for line in body.splitlines():
        match = re.match(r"^#\s+(.+?)\s*$", line)
        if match:
            title = match.group(1).strip()
            break

    if not title:
        title = path.stem.replace("-", " ").title()
    if not description:
        description = title
    return title, description


def installed_rule_path(path: Path) -> str:
    return f"~/.agents/rules/{path.name}"


def render_rules() -> str:
    if not RULES_DIR.exists():
        return "_No shared rules are currently defined._\n"

    lines: list[str] = []
    for path in sorted(RULES_DIR.glob("*.md")):
        title, description = parse_rule(path)
        rel = path.relative_to(ROOT)
        lines.append(f"- `{rel}` ({installed_rule_path(path)}): {title} - {description}")

    if not lines:
        return "_No shared rules are currently defined._\n"
    return "\n".join(lines) + "\n"


def main() -> None:
    template = TEMPLATE.read_text(encoding="utf-8")
    if BEGIN not in template or END not in template:
        raise SystemExit(f"{TEMPLATE} must contain generated rules markers")

    before, rest = template.split(BEGIN, 1)
    _, after = rest.split(END, 1)
    output = before + BEGIN + "\n" + render_rules() + END + after

    if OUTPUT.exists() and OUTPUT.read_text(encoding="utf-8") == output:
        return
    OUTPUT.write_text(output, encoding="utf-8")


if __name__ == "__main__":
    main()

File: crates/xtask/src/render_agents.rs
use std::fs;
use std::path::Path;

use anyhow::{Context, Result, bail};
use regex::Regex;

const BEGIN: &str = "<!-- BEGIN GENERATED RULES -->";
const END: &str = "<!-- END GENERATED RULES -->";

pub fn run() -> Result<()> {
    let root = crate::repo_root();
    let template_path = root.join("AGENTS.template.md");
    let output_path = root.join("GLOBAL_AGENTS.md");
    let rules_dir = root.join("rules");

    let template = fs::read_to_string(&template_path)
        .with_context(|| format!("read {}", template_path.display()))?;
    let begin_idx = template
        .find(BEGIN)
        .with_context(|| format!("{} missing {BEGIN}", template_path.display()))?;
    let end_idx = template
        .find(END)
        .with_context(|| format!("{} missing {END}", template_path.display()))?;
    if end_idx < begin_idx {
        bail!("{} markers are out of order", template_path.display());
    }

    let before = &template[..begin_idx];
    let after = &template[end_idx + END.len()..];
    let rules_block = render_rules(&rules_dir, &root)?;
    let new_output = format!("{before}{BEGIN}\n{rules_block}{END}{after}");

    if let Ok(existing) = fs::read_to_string(&output_path)
        && existing == new_output
    {
        return Ok(());
    }
    fs::write(&output_path, new_output)
        .with_context(|| format!("write {}", output_path.display()))?;
    Ok(())
}

fn render_rules(rules_dir: &Path, root: &Path) -> Result<String> {
    if !rules_dir.exists() {
        return Ok("_No shared rules are currently defined._\n".to_string());
    }
    let mut entries: Vec<_> = fs::read_dir(rules_dir)
        .with_context(|| format!("read_dir {}", rules_dir.display()))?
        .filter_map(|e| e.ok())
        .map(|e| e.path())
        .filter(|p| p.extension().map(|e| e == "md").unwrap_or(false))
        .collect();
    entries.sort();

    if entries.is_empty() {
        return Ok("_No shared rules are currently defined._\n".to_string());
    }

    let mut lines = Vec::with_capacity(entries.len());
    for path in entries {
        let (title, description) = parse_rule(&path)?;
        let rel = path
            .strip_prefix(root)
            .unwrap_or(&path)
            .display()
            .to_string();
        let name = path.file_name().unwrap().to_string_lossy();
        lines.push(format!(
            "- `{rel}` (~/.agents/rules/{name}): {title} - {description}"
        ));
    }
    Ok(lines.join("\n") + "\n")
}

fn parse_rule(path: &Path) -> Result<(String, String)> {
    let text = fs::read_to_string(path).with_context(|| format!("read {}", path.display()))?;
    let (description, body) = if let Some(rest) = text.strip_prefix("---\n") {
        if let Some(end) = rest.find("---\n") {
            let frontmatter = &rest[..end];
            let body = &rest[end + 4..];
            let mut desc = String::new();
            for line in frontmatter.lines() {
                if let Some((key, value)) = line.split_once(':') {
                    let key = key.trim();
                    if key == "description" || key == "summary" {
                        desc = value.trim().trim_matches(['"', '\'']).to_string();
                        break;
                    }
                }
            }
            (desc, body.to_string())
        } else {
            (String::new(), text.clone())
        }
    } else {
        (String::new(), text.clone())
    };

    let header = Regex::new(r"^#\s+(.+?)\s*$").unwrap();
    let mut title = String::new();
    for line in body.lines() {
        if let Some(caps) = header.captures(line) {
            title = caps[1].trim().to_string();
            break;
        }
    }
    if title.is_empty() {
        let stem = path.file_stem().unwrap().to_string_lossy();
        title = stem
            .split('-')
            .map(|w| {
                let mut chars = w.chars();
                match chars.next() {
                    Some(c) => c.to_uppercase().collect::<String>() + chars.as_str(),
                    None => String::new(),
                }
            })
            .collect::<Vec<_>>()
            .join(" ");
    }
    let description = if description.is_empty() {
        title.clone()
    } else {
        description
    };
    Ok((title, description))
}
